### PR TITLE
Override Read(Span), ReadAsync(Memory), Write(ROS), WriteAsync(ROM) in streams

### LINKED ISSUE #1570: Override Read(Span), ReadAsync(Memory), Write(ROS), WriteAsync(ROM) in streams
Available in NETSTANDARD2_1 and NET5_0_OR_GREATER, the base class functions in Stream for backwards compatibility rent an array and copy the span to it and then call the old array based functions.

As an optimization `FtpDataStream` and `FtpSocketStream` can override these functions and pass the call directly onto the wrapped stream.

```
public override void Write(ReadOnlySpan<byte> buffer) {
    if (BaseStream == null) {
        return;
     }

    BaseStream.Write(buffer);
    m_lastActivity = DateTime.UtcNow;
}
```

Like how the old FtpClient has been changed https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Requests/src/System/Net/FtpDataStream.cs#L231

### PR DESCRIPTION
**Pull request by bbowyersmyth**
Available in NETSTANDARD2_1 and NET5_0_OR_GREATER, the base class functions in Stream for backwards compatibility rent an array and copy the span to it and then call the old array based functions.

As an optimization FtpDataStream and FtpSocketStream can override these functions and pass the call directly onto the wrapped stream.

Closes #1570

### CODE REVIEW

**Comment by robinrodricks** [comment]
The changes look good. But are these new methods automatically used by the .NET Framework or the Stream system? As we are not invoking it directly.

File: FluentFTP/Streams/FtpSocketStream.cs [hunk 1]
```diff
@@ -591,6 +607,44 @@ public class FtpSocketStream : Stream, IDisposable {
...
+		/// <param name="token">The <see cref="CancellationToken"/> for this task</param>
+		/// <returns>The amount of bytes read from the stream</returns>
+		public override async ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken token) {
+			if (BaseStream == null) {
+				return 0;
+			}
+
+			m_lastActivity = DateTime.UtcNow;
+			using (var cts = CancellationTokenSource.CreateLinkedTokenSource(token)) {
+				cts.CancelAfter(ReadTimeout);
```
Comment (robinrodricks) [inline FluentFTP/Streams/FtpSocketStream.cs:624]: what's all this about?

**Comment by bbowyersmyth** [comment]
@robinrodricks Yes stream and other places in system will now prefer to use the new overloads if they don't already have an array. For example `FileStream.CopyToAsync` uses the memory overload https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/FileStreamHelpers.Windows.cs#L299C39-L299C49

So they may already be getting called. Will know if we can put a break point in there when the tests are run. Otherwise we will need to add more.

File: FluentFTP/Streams/FtpSocketStream.cs (hunk 1 above)
Comment (bbowyersmyth) [reply FluentFTP/Streams/FtpSocketStream.cs:624]: Same code as the existing ReadAsync above. It will be setting a timer to timeout the async call.

**Comment by FanDjango** [comment]
I have also been investigating bypassing the BUF copying inherent in the FtpDataStream and FtpSocketStream on the way to (or from) network stream.

Newer .NET versions give one this option:

**Prefer the memory-based overloads of ReadAsync/WriteAsync methods in stream-based classes**

For example: `await s.ReadAsync(buffer.AsMemory(0, buffer.Length), token);`

Would of course be a code change and another set of #IF/#ENDIF brackets.

**Comment by bbowyersmyth** [comment]
Added 2 more UploadDownload tests as the existing ones weren't using the streams directly.
Fixed a couple of compiler warnings in the file also.

**Comment by robinrodricks** [comment]
This looks interesting. I'm merging it for now. If we have serious issues arising from this change, we can revert it.

Thanks for your contribution!

**Pull request merged**